### PR TITLE
Synchronized vectorlist

### PR DESCRIPTION
**Pull request by christopherlagerhult**
Added abstract class _List
Vectorlist class now inherits _List
Added class SynchronizationList

### CODE REVIEW

**Comment by marcusnaslund** [comment]
Note: You have conflicts, so be sure to rebase before you develop much more.

File: source/concurrent/SynchronizedList.ooc [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	}
+	removeAt: override func (index: Int) {
+		this _mutex lock()
+		this _backend removeAt(index)
+		this _mutex unlock()
+	}
+	clear: override func {
+		this _mutex lock()
+		this _backend clear()
+		this _mutex unlock()
```
Comment (marcusnaslund) [inline source/concurrent/SynchronizedList.ooc]: `Mutex` has a `with` method which does this. I don't know if `this _mutex with(this _backend clear)` would work, though. Maybe.

File: test/concurrent/SynchronizedListTest.ooc [hunk 2]
```diff
@@ -0,0 +1,328 @@
+use ooc-base
+use ooc-concurrent
+use ooc-unit
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: These should not have the `ooc-` prefix anymore.

File: source/concurrent/SynchronizedList.ooc [hunk 3]
```diff
@@ -0,0 +1,143 @@
+import structs/[VectorList, List]
```
Comment (marcusnaslund) [inline source/concurrent/SynchronizedList.ooc]: You do not need to import `VectorList`. Also, you should add `List` to `sdk.use`, and you won't need to import it anywhere either.

File: test/concurrent/SynchronizedListTest.ooc [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+				expect(lastValue, is equal to(removedValue))
+			}
+			expect(list count, is equal to(900))
+			for (value in 0 .. 100) {
+				lastValue := list[list count - 1]
+				removedValue := list remove(list count - 1)
+				expect(lastValue, is equal to(removedValue))
+			}
+			expect(list count, is equal to(800))
+			for (value in 1 .. 800) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+			expect(list count, is equal to(6))
+			expect(appendList count, is equal to(3))
+			list free()
+			appendList free()
+		})
+		this add("single thread - insert", func {
+			list := SynchronizedList<Int> new()
+			list add(0) . add(2) . add(4) . add(6) . add(8)
+			list insert(1, 1) . insert(3, 3) . insert(5, 5) . insert(7, 7) . insert(9, 9)
+			for (value in 0 .. 10) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+				expect(list[value], is equal to(value))
+			}
+			expect(list count, is equal to(10))
+			list free()
+		})
+		this add("single thread - removeAt", func {
+			list := SynchronizedList<Int> new()
+			list add(0) . add(1) . add(2) . add(3) . add(4)
+			for (value in 0 .. 5) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+				expect(list[value], is equal to(value))
+			}
+			expect(list count, is equal to(5))
+			list free()
+		})
+		this add("Multi thread - clear", func {
+			list := SynchronizedList<Int> new()
+			expect(list count, is equal to(0))
+			for (value in 0 .. 1000) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+				list add(value)
+			}
+			list clear()
+			expect(list count, is equal to(0))
+			list free()
+		})
+		this add("single thread - reverse", func {
+			list := SynchronizedList<Int> new()
+			for (value in 0 .. 1000) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+			}
+			expect(reversedList count, is equal to(1000))
+			list free()
+			reversedList free()
+		})
+		this add("single thread - sort", func {
+			list := SynchronizedList<Int> new()
+			list add(Int minimumValue) . add(Int maximumValue) . add(0) . add(-10) . add(10)
+			list sort(|first, second| first < second)
+			for (value in 0 .. list count - 1) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+				expect(list[value] == copiedList[value] - 1)
+			}
+			list free()
+			copiedList free()
+		})
+		this add("single thread - modify", func {
+			list := SynchronizedList<Int> new()
+			list add(-5) . add(-4) . add(-3) . add(-2) . add(-1)
+			list modify(|value| value += 5)
+			for (value in 0 .. list count) {
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You don't need these `{` `}` brackets.

File: test/concurrent/SynchronizedListTest.ooc [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+				expect(list[value], is greater than(list[value + 1]))
+			}
+			list free()
+		})
+		this add("single thread - copy", func {
+			list := SynchronizedList<Int> new()
+			list add(-5) . add(-4) . add(-3) . add(-2) . add(-1)
+			copiedList := list copy()
+			for (value in 0 .. list count) {
+				expect(list[value] == copiedList[value])
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: Should be `expect(list[value], is equal to(copiedList[value])`

File: test/concurrent/SynchronizedListTest.ooc [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+			result free()
+		})
+		this add("single thread - getElements", func {
+			list := SynchronizedList<Int> new()
+			list add(-5) . add(-4) . add(-3) . add(-2) . add(-1)
+			indicesList := SynchronizedList<Int> new() . add(2) . add(4)
+			result := list getElements(indicesList)
+			expect(result count == 2)
+			expect(result[0] == list[2])
+			expect(result[1] == list[4])
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You should use `is equal to` here and above.

File: test/concurrent/SynchronizedListTest.ooc [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+			sum := list fold(Int, |v1, v2| v1 + v2, 0)
+			expect(sum, is equal to(6))
+		})
+		this add("single thread - getFirstElement", func {
+			list := SynchronizedList<Int> new()
+			list add(-5) . add(-4) . add(-3) . add(-2) . add(-1)
+			result := list getFirstElements(2)
+			expect(result count == 2)
+			expect(result[0] == list[0])
+			expect(result[1] == list[1])
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You should use `is equal to` here and above.

File: test/concurrent/SynchronizedListTest.ooc [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+			list free()
+			sliceInto free()
+		})
+		this add("single thread - operators", func {
+			list := SynchronizedList<Int> new()
+			list add(-5) . add(-4) . add(-3) . add(-2) . add(-1)
+			oldValue := list[0]
+			list[0] = -6
+			expect(oldValue == -5)
+			expect(list[0] == -6)
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You should use `is equal to` here and above.

File: test/concurrent/SynchronizedListTest.ooc [hunk 15]
```diff
@@ -0,0 +1,328 @@
...
+				expect(value >= 0 && value < countPerThread)
+			}
+		}
+		for (i in 0 .. numberOfThreads) {
+			threads[i] = Thread new(job)
+			threads[i] start()
+		}
+		for (i in 0 .. numberOfThreads) {
+			threads[i] wait()
+			threads[i] free()
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: You can do: `threads[i] wait() . free()` here and in other places in this test.
Also, 
```ooc
threads[i] = Thread new(job)
threads[i] start()
```
can probably be written `threads[i] = Thread new(job) . start()`

File: test/concurrent/SynchronizedListTest.ooc [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+		countPerThread := 10_000
+		list := SynchronizedList<Int> new()
+		threads := Thread[numberOfThreads] new()
+		job := func {
+			for (i in 0 .. countPerThread)
+				list add(i)
+		}
+		for (i in 0 .. numberOfThreads) {
+			threads[i] = Thread new(job)
+			threads[i] start()
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: Also, 
```ooc
threads[i] = Thread new(job)
threads[i] start()
```
can probably be written `threads[i] = Thread new(job) . start()`

File: test/concurrent/SynchronizedListTest.ooc [hunk 17]
```diff
@@ -0,0 +1,328 @@
...
+		for (i in 0 .. numberOfThreads) {
+			threads[i] wait()
+			threads[i] free()
+		}
+		expect(list count, is equal to(numberOfThreads * countPerThread))
+		(job as Closure) free()
+		job = func {
+			for (i in 0 .. countPerThread) {
+				value: Cell<Int>
+				value = list remove(0)
```
Comment (marcusnaslund) [inline test/concurrent/SynchronizedListTest.ooc]: `value := list remove(0)`?

File: source/math/IntVectorList.ooc [hunk 18]
```diff
@@ -35,7 +35,7 @@ IntVectorList: class extends VectorList<Int> {
 			this add(value)
 	}
 	copy: func -> This {
-		this as VectorList<Int> copy() as This
+		super() as This
```
Comment (marcusnaslund) [inline source/math/IntVectorList.ooc]: Does this work in `FloatComplexVectorList copy` and `FloatVectorList copy` too?

File: source/sdk/structs/VectorList.ooc [hunk 19]
```diff
@@ -15,14 +15,11 @@
 * along with this software. If not, see <http://www.gnu.org/licenses/>.
 */
 
-import structs/Vector
+import structs/[Vector, List]
```
Comment (marcusnaslund) [inline source/sdk/structs/VectorList.ooc]: You don't need to import `List`

**Comment by marcusnaslund** [comment]
`GCC` generates a lot of warnings when building both `VectorList` and `SynchronizedList`. That will have to be fixed before this gets the green flag.

**Comment by marcusnaslund** [comment]
I can't run this with valgrind either:
`rock -x && ./test.sh concurrent/SynchronizedList && ./tools/memcheck.sh`
which suggests some there's a race condition somewhere.

File: source/sdk/structs/VectorList.ooc (hunk 19 above)
Comment (marcusnaslund) [inline source/sdk/structs/VectorList.ooc]: While you're at it, add `Vector` to `sdk.use` just as you did for `List`, and you can remove this entire line.

**Comment by sebastianbaginski** [comment]
> GCC generates a lot of warnings

@marcusnaslund : maybe separating the `VectorList extends _List` part into other PR could be a good idea ?

**Comment by marcusnaslund** [comment]
> maybe separating the VectorList extends _List part into other PR could be a good idea ?

Should have been done like that, yes, but I think we've solved that problem.

**Comment by marcusnaslund** [comment]
Note: This PR no longer passes tests because our CI-service from now on fails on all gcc warnings.

File: source/concurrent/SynchronizedList.ooc [hunk 20]
```diff
@@ -0,0 +1,142 @@
+import threading/Mutex
+
+SynchronizedList: class <T> extends List<T> {
+	_mutex := Mutex new()
+	_backend : VectorList<T>
```
Comment (marcusnaslund) [inline source/concurrent/SynchronizedList.ooc]: You don't need space before this colon

**Comment by marcusnaslund** [comment]
Welcome to the "PR merged" club. :)

**Pull request merged**